com.unity.multiplayer.mlapi/Runtime/Core/SnapshotSystem.cs

```diff
@@ -0,0 +1,412 @@
+    internal struct Key
+    {
+        public ulong NetworkObjectId; // the NetworkObjectId of the owning GameObject
+        public ushort BehaviourIndex; // the index of the behaviour in this GameObject
+        public ushort VariableIndex; // the index of the variable in this NetworkBehaviour
+    }
+
+    // Index for a NetworkVariable in our table of variables
+    // Store when a variable was written and where the variable is serialized
+    internal struct Entry
+    {
+        public Key Key;
+        public ushort TickWritten; // the network tick at which this variable was set
+        public ushort Position; // the offset in our Buffer
+        public ushort Length; // the length of the data in Buffer
+        public bool Fresh; // indicates entries that were just received
+
+        public const int NotFound = -1;
+    }
+
+    // A table of NetworkVariables that constitutes a Snapshot.
+    // Stores serialized NetworkVariables
+    // todo --M1--
+    // The EntryBlock will change for M1b with memory management, instead of just Beg, there will be data structure
+    // around available buffer, etc.
+    internal class EntryBlock
+    {
+        // todo --M1-- functionality to grow these will be needed in a later milestone
+        private const int k_MaxVariables = 64;
+        private const int k_BufferSize = 20000;
+
+        public byte[] Buffer = new byte[k_BufferSize];
+        public int Beg = 0; // todo: clarify usage. Right now, this is the beginning of the _free_ space.
+        public int End = 0;
+
+        public Entry[] Entries = new Entry[k_MaxVariables];
+        public int LastEntry = 0;
+        public MemoryStream Stream;
+
+        /// <summary>
+        /// Constructor
+        /// Allocated a MemoryStream to be reused for this EntryBlock
+        /// </summary>
+        public EntryBlock()
+        {
+            Stream = new MemoryStream(Buffer, 0, k_BufferSize);
+        }
+
+        // todo --M1--
+        // Find will change to be efficient in a future milestone
+        /// <summary>
+        /// Finds the position of a given NetworkVariable, given its key
+        /// </summary>
+        /// <param name="key">The key we're looking for</param>
+        public int Find(Key key)
+        {
+            for (int i = 0; i < LastEntry; i++)
+            {
+                if (Entries[i].Key.NetworkObjectId == key.NetworkObjectId &&
+                    Entries[i].Key.BehaviourIndex == key.BehaviourIndex &&
+                    Entries[i].Key.VariableIndex == key.VariableIndex)
+                {
+                    return i;
+                }
+            }
+
+            return Entry.NotFound;
+        }
+
+        /// <summary>
+        /// Adds an entry in the table for a new key
+        /// </summary>
+        // todo: change the params to take a Key instead
+        public int AddEntry(ulong networkObjectId, int behaviourIndex, int variableIndex)
```

Comment:
Is there any moment where an allocation could be made without an AddEntry? I'm curious why the two are separate. If I do an AddEntry alone without an allocation, I could leave an entry with position and length set to 0 which is kind of invalid no?

Reply:
I wanted to have separate function for the two separate aspect: allocating the index versus allocating the memory. In fact, position and length set to 0 are a valid combination. It's not a very useful one, as you can store a total of 0 bytes there. But it is still valid, it could (does?) represent "no allocated memory for this entry".

Reply:
what would be the use for an unallocated entry? Be able to represent null values for example?